scylla/src/transport/connection_pool.rs

```diff
@@ -80,6 +90,61 @@ enum PoolConnections {
     },
 }
 
+struct ConnectionVectorWrapper<'a>(&'a Vec<Arc<Connection>>);
+impl std::fmt::Display for ConnectionVectorWrapper<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            self.0
+                .iter()
+                .map(|conn| conn.get_connect_address().to_string())
+                .collect::<Vec<String>>()
+                .join(",")
+        )
+    }
```

Comment:
I think you can reduce the number of allocations in your `impl Display`s. Take a look at the docs for `Formatter` ([link](https://doc.rust-lang.org/std/fmt/struct.Formatter.html)), especially the `debug_xyz` family of methods. You can use them to print lists, sets, maps, etc.

For example, this one could be rewritten as:

```rust
f.debug_list()
    .entries(self.0.iter().map(|conn| conn.get_connect_address()))
    .finish()
```

This will change the formatting a bit, there will be additional brackets and `Debug` trait will be used to print IPs. If that's not acceptable, we can stay with the current formatting or you could open-code a for loop and use `write!(f, "{}", conn.get_connect_address())` to print the addresses (taking care of separator commas ofc).

This could be done for other impls, too.

Reply:
Yeah, I was tempted to do that, but I thought that maybe it's considered bad taste to use debug formatting in order to implement non-debug string formatting. But you're right that it makes the implementation considerably simpler, so I'll do that, thx.

Reply:
I went a little further and just implemented `Debug` for these wrappers - their output is pretty much only used for debugging so it fits the purpose, and now I don't have a moral hangover of implementing `Display` by relying on `Debug`.